### PR TITLE
Refactor block reward application and tracing

### PR DESCRIPTION
**Pull request by andresilva**
Refactors block reward application to use a common helper function. Add support for tracing different kinds of rewards.

### CODE REVIEW

File: ethcore/src/engines/authority_round/mod.rs [hunk 1]
```diff
@@ -1027,10 +1029,11 @@ impl Engine<EthereumMachine> for AuthorityRound {
 					result.map_err(|e| format!("{}", e))
 				};
 
-				c.reward(&benefactors, &mut call)?
+				let rewards = c.reward(&benefactors, &mut call)?;
+				rewards.into_iter().map(|(author, amount)| (author, RewardKind::External, amount)).collect()
```
Comment (andresilva) [inline ethcore/src/engines/authority_round/mod.rs]: Unfortunately when using the block reward contract all rewards must be tagged as `External`, there's no way to appropriately map to benefactors list that was passed to the contract.

File: ethcore/src/trace/types/trace.rs [hunk 2]
```diff
@@ -141,13 +141,19 @@ pub enum RewardType {
 	Block,
 	/// Uncle
 	Uncle,
+	/// Empty step (AuthorityRound)
```
Comment (andresilva) [inline ethcore/src/trace/types/trace.rs]: Should we remove this type from tracing and instead use `block_reward::RewardKind`?

File: ethcore/src/engines/block_reward.rs [hunk 3]
```diff
@@ -14,13 +14,17 @@
 // You should have received a copy of the GNU General Public License
 // along with Parity.  If not, see <http://www.gnu.org/licenses/>.
 
+//! A module with types for declaring block rewards and a client interface for interacting with a
```
Comment (andresilva) [inline ethcore/src/engines/block_reward.rs]: I'm not sure `engines` is the appropriate module for `block_reward`. Any thoughts?

File: ethcore/src/trace/types/trace.rs (hunk 2 above)
Comment (tomusdrw) [reply ethcore/src/trace/types/trace.rs]: Sounds ok for me.

**Pull request merged**